### PR TITLE
feat(test): add chroot as integration test platform (#709)

### LINKED ISSUE #709: Create new Provider 'chroot' for Integration Tests
**What would you like to be added**:
A new provider `chroot` should be created for integration tests. This makes it possible for everyone to execute unit tests as integration tests based on the final artifact without any further cloud provider access.

This feature is the future base for the transition (https://github.com/gardenlinux/gardenlinux/issues/676) from unit tests to integration tests. These tests will be executed by the `chroot` platform as unit tests.

**Goal:** All tests via pytest

**Why is this needed**:
 * Tests are executed with pytest
 * This ensures that everyone can execute integration
 * No charges for cloud provider usage
 * Shared tests between unit- & integration tests (based on chroot) that may also be used by other platforms
 * Solves dependency from: https://github.com/gardenlinux/gardenlinux/issues/676

### PR DESCRIPTION
**Pull request by gyptazy**
Add 'chroot' as a new integration platform to perform tests.
  * New 'chroot' integration platform
  * Adjusted 'RemoteClient' to support 'chroot'
  * Adjusted 'test_gardenlinux.py' to support in/exclude for 'chroot' tests

**How to categorize this PR?**
/kind enhancement
/area os
/os garden-linux

**What this PR does / why we need it**:
This PR adds a new platform provider for integration tests. This allows us to perform tests inside a `chroot`. Next to the new `KVM` this allows users to run integration (and future unit) tests without any cloud platform like GCP, AWS,....  The new `chroot` platform is also the base for the migration (#676) from unit tests to a common Pytest base.

**Which issue(s) this PR fixes**:
Fixes #709

**Special notes for your reviewer**:
The provided `test_config.yaml` can be used for testing this feature without any needed changes. Make sure to have a matching tar ball - this can be created by running:
```
make kvm-dev
```

`test_config.yaml` can be removed before merging.

**Release note**:

NONE
```feature user

```

### CODE REVIEW

File: tests/integration/chroot.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        except tarfile.CompressionError:
+            logger.error("Used compression is not supported.")
+        except tarfile.HeaderError:
+            logger.error("Malformed tar header information.")
+        logger.info("Unarchived image {image} to {tmp_dir}".format(
+          image=image, tmp_dir=tmp_dir))
+        return tmp_dir
+
+
+    def _generate_ssh_key(self):
```
Comment (MrBatschner) [inline tests/integration/chroot.py:149]: This functionality is needed everywhere. There is a function `RemoteClient.generate_key_pair()` can you reuse that one?

**Review by MrBatschner — CHANGES_REQUESTED** [review]
See above.

Note that there are huge changes to the integration test framework in the branch `testpipeline`, but since you were first committing your changes, I will integrate them into the `testpipeline` branch.

File: tests/integration/chroot.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            logger.info("Copied sshd_config_integration_tests to: {dir}".format(
+              dir=chroot_sshd_cfg_dir))
+        except OSError:
+            logger.error("Could not copy sshd_config_integration_tests to {dir}".format(
+              dir=chroot_sshd_cfg_dir))
+        # Copy ssh / authorized_keys file for root user
+        chroot_root_dir = tmp_dir + "/root/"
+        chroot_ssh_authorized_keys = chroot_root_dir + ".ssh/authorized_keys"
+        ssh_authorized_keys = "/tmp/authorized_keys"
+        self._create_dir(chroot_root_dir+".ssh", 0o600)
```
Comment (nanory) [inline tests/integration/chroot.py:225]: I would use the following command here:
```
access = 0o600
os.mkdir(path, access)
```
Then, there is no need for this helper function

File: tests/integration/chroot.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        """ Wait for defined SSH port to become ready in chroot environment """
+        port = self.config["port"]
+        logger.info("Waiting for SSHD in chroot to be ready on tcp/{port}...".format(
+          port=port))
+        cmd_kvm = "ssh-keyscan -p {port} localhost".format(
+          port=port)
+
+        rc = 1
+        while rc != 0:
+            p = subprocess.Popen([cmd_kvm], shell=True, stdout=subprocess.PIPE, \
```
Comment (nanory) [inline tests/integration/chroot.py:258]: Why not `subprocess.run` ?

File: tests/integration/chroot.py (hunk 2 above)
Comment (nanory) [inline tests/integration/chroot.py:222]: I would use the following command here:
```
access = 0o600
os.mkdir(path, access)
```
Then, there is no need for this helper function

File: tests/integration/chroot.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        p = subprocess.Popen([proc_exec], shell=True)
+        logger.info("Started SSHD in chroot environment.")
+
+
+    def _wait_sshd(self):
+        """ Wait for defined SSH port to become ready in chroot environment """
+        port = self.config["port"]
+        logger.info("Waiting for SSHD in chroot to be ready on tcp/{port}...".format(
+          port=port))
+        cmd_kvm = "ssh-keyscan -p {port} localhost".format(
```
Comment (nanory) [inline tests/integration/chroot.py:253]: I would rename this variable from `cmd_kvm` to `cmd`

File: tests/integration/chroot.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+            logger.warning("'garden-epoch' already present. This will be set to 'dev', now.")
+        try:
+            with open(chroot_epoch, 'w') as f:
+                f.write("dev")
+            logger.info("Wrote garden-epoch to chroot in {tmp_dir}".format(
+              tmp_dir=tmp_dir))
+        except IOError:
+            logger.error("Could not write garden-epoch to chroot in {tmp_dir}".format(
+              tmp_dir=tmp_dir))
+        # Generate SSH hostkeys for chroot
```
Comment (nanory) [inline tests/integration/chroot.py:175]: For better readability, I would suggest to add a newline in front of each comment section. Similar to the `_validate` method

File: tests/integration/chroot.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        logger.info("SSH key for integration tests generated.")
+
+
+    def _adjust_chroot(self, tmp_dir):
+        """ Adjust chroot to make it usable """
+        logger.info("Adjusting chroot environment")
+        # Adjust garden-epoch to dev
+        chroot_epoch = tmp_dir + "/" + "garden-epoch"
+        chroot_epoch_bool = os.path.exists(chroot_epoch)
+        if chroot_epoch_bool:
```
Comment (nanory) [inline tests/integration/chroot.py:165]: `chroot_epoch_bool` is only used in this if statement. If there is no other need, I would probably define it like this:
```
if os.path.exists(chroot_epoch):
```

File: tests/integration/chroot.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+        cmd_kvm = "ssh-keyscan -p {port} localhost".format(
+          port=port)
+
+        rc = 1
+        while rc != 0:
+            p = subprocess.Popen([cmd_kvm], shell=True, stdout=subprocess.PIPE, \
+              stderr=subprocess.STDOUT)
+            output = p.stdout.read()
+            output, error = p.communicate()
+            rc = p.returncode
```
Comment (nanory) [inline tests/integration/chroot.py:262]: We should probably harden this while loop. Instead of keeping this running endlessly, we should check the return code or at least the output for cases that are not related for an `sshd` that is still starting (e.g connection issues).

Alternatively, we should add a retry counter and fail if it reached a specific limit.

File: tests/integration/chroot.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+
+
+    def __init__(self, config):
+        """ Creating self config """
+        # Define self.config
+        self.config = config
+        # Perform basic validation
+        self._validate()
+        # Unarchive defined tar ball
+        tmp_dir = self._unarchive_image()
```
Comment (nanory) [inline tests/integration/chroot.py:73]: In order to make it easier to understand what the content of this tmp_dir is, I would suggest to rename this variable and the corresponding parameters from `tmp_dir` to `rootfs`. It makes it easier for someone who already implemented chroot tests in `garden-build` since the variable there is called `rootfs` as well.

**Review by nanory — CHANGES_REQUESTED** [review]

File: tests/integration/chroot.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+        # while we may perform our integration tests
+        proc_exec = "{chroot_bin} {chroot_env} {chroot_cmd}".format(
+          chroot_bin=gl_chroot_bin,
+          chroot_env=chroot,
+          chroot_cmd=chroot_cmd)
+        p = subprocess.Popen([proc_exec], shell=True)
+        logger.info("Started SSHD in chroot environment.")
+
+
+    def _wait_sshd(self):
```
Comment (nanory) [inline tests/integration/chroot.py:248]: This method is used by `kvm` and `chroot`, no?
It may makes sense to move this method to another module. Maybe `sshclient.py`. What to you think?

File: tests/integration/chroot.py (hunk 1 above)
Comment (gyptazy) [reply tests/integration/chroot.py:149]: Okay, done. But I think `generate_key_pair` should be more fail safe. While playing around with it this function has overwritten my private key. Luckily not a big deal in my case but I think there must be a check to validate if the given file is already present and not to overwrite it by default... If some users want to mount their prod key, this can be very risky...

```
$> md5sum /tmp/ssh_priv_key >> /tmp/foo && pytest --iaas=chroot --configfile=test_config.yaml integration/ && md5sum /tmp/ssh_priv_key >> /tmp/foo && cat /tmp/foo
99a5747c2e338b2a94dc5c58b7a369f9  /tmp/ssh_priv_key
e5f57dd4e229e62e222e095053d7963d  /tmp/ssh_priv_key
```

For `chroot` this isn't a big deal (there's only auto generation available) but it may become one for other cloud platforms. Maybe we just raise a issue for this...

File: tests/integration/chroot.py (hunk 2 above)
Comment (gyptazy) [reply tests/integration/chroot.py:222]: There is no need fore, but IMHO it looks much better. Furthermore, there'd be more lines for handling errors and umask.

File: tests/integration/chroot.py (hunk 3 above)
Comment (gyptazy) [reply tests/integration/chroot.py:258]: yep, does not need to run as bg.

File: tests/integration/chroot.py (hunk 5 above)
Comment (gyptazy) [reply tests/integration/chroot.py:175]: Probably more a question regarding the `Style Guide` - while looking at other files it is mostly used this way.

File: tests/integration/chroot.py (hunk 8 above)
Comment (gyptazy) [reply tests/integration/chroot.py:73]: Makes sense.

File: tests/integration/chroot.py (hunk 6 above)
Comment (gyptazy) [reply tests/integration/chroot.py:165]: ok

File: tests/integration/chroot.py (hunk 9 above)
Comment (gyptazy) [reply tests/integration/chroot.py:248]: Ok, but I think we should raise a new issue for this - so we can adjust both ones directly. However, we may also adjust https://github.com/gardenlinux/gardenlinux/pull/712#discussion_r833204784.

File: tests/integration/chroot.py (hunk 7 above)
Comment (gyptazy) [reply tests/integration/chroot.py:262]: Could be solves with https://github.com/gardenlinux/gardenlinux/pull/712#discussion_r833259181

File: tests/integration/chroot.py (hunk 4 above)
Comment (nanory) [reply tests/integration/chroot.py]: As discussed in comment https://github.com/gardenlinux/gardenlinux/pull/712#discussion_r833259181, since we will move this method in its own module, I would say we need to adjust this cmd anyways, so if `cmd` is not feasible enough, I would rather call this command `keyscan_cmd` or `ssh_cmd`/`cmd_ssh`

**Review by nanory — APPROVED** [review]
